crates/biome_js_analyze/src/semantic_analyzers/nursery/no_global_assign.rs

```diff
@@ -0,0 +1,141 @@
+use crate::globals::browser::BROWSER;
+use crate::globals::node::NODE;
+use crate::globals::runtime::{BUILTIN, ES_2021};
+
+use crate::semantic_services::Semantic;
+use biome_analyze::{context::RuleContext, declare_rule, Rule, RuleDiagnostic};
+use biome_console::markup;
+use biome_js_syntax::AnyJsExpression;
+use biome_rowan::AstNode;
+
+declare_rule! {
+    /// Disallow assignments to native objects or read-only global variables.
+    ///
+    /// JavaScript environments contain numerous built-in global variables, such as 'window' in browsers and 'process' in Node.js.
+    /// Assigning values to these global variables can lead to problems, as it might overwrite essential functionality.
+    ///
+    /// Source: https://eslint.org/docs/latest/rules/no-global-assign
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// Object = null;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// window = 1;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// undefined = 1;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// /*global a:readonly*/
+    /// a = 1;
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```js
+    /// const a = 1;
+    /// let b = 1;
+    /// b = 100;
+    ///
+    /// onload = function() {};
+    ///
+    /// /*global a:writable*/
+    /// a = 100;
+    /// ```
+    ///
+    pub(crate) NoGlobalAssign {
+        version: "next",
+        name: "noGlobalAssign",
+        recommended: false,
+    }
+}
+
+impl Rule for NoGlobalAssign {
+    type Query = Semantic<AnyJsExpression>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let expr = match node {
+            AnyJsExpression::JsAssignmentExpression(expr) => expr,
+            _ => return None,
+        };
```

Comment:
Instead of checking every assignment, we could retrieve global references from the semantic model, check if the reference is a write and is among known global variables (BROWSER, BUILTIN, ...).

We could then query `SematicServices`, and return a vector of text ranges (ranges of the references to report)

This could look something like this:

```suggestion
    type Query = SemanticServices;
    type State = TextRange;
    type Signals = Vec<Self::State>;
    type Options = ();

    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
        let global_refs = ctx.query().all_unresolved_references();
        let result = vec![];
        for global_ref in global_refs {
            let is_write = global_ref.syntax().kind() == JsSyntaxKind::JS_IDENTIFIER_ASSIGNEMENT;
            if is_write {
                let is_global_var = ...
                if is_global_var {
                    result.push(global_ref.range());
                }
            }
        }
        
```